lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,276 @@
+        Value add = op.getBody()->getArgument(tileOffsets[oldArgIndex] + 1);
+        Value mul;
+        for (int j = tileOffsets[oldArgIndex] + 2;
+             j <= tileOffsets[oldArgIndex + 1]; j++) {
+          Value upperBound = b.create<arith::ConstantIndexOp>(
+              op.getLoc(), op.getStaticUpperBound()[j]);
+          mul = b.create<arith::MulIOp>(op.getLoc(), b.getIndexType(), add,
+                                        upperBound);
+          add = b.create<arith::AddIOp>(op.getLoc(), b.getIndexType(), mul,
+                                        op.getBody()->getArgument(j));
+        }
+        oper->setOperand(operandIndex, add);
+      }
+      operandIndex++;
+    }
+  }
+
+  SmallVector<ArrayRef<int64_t>> originalShapes{mShape, nShape, kShape};
+  SmallVector<SmallVector<ArrayRef<unsigned>>> tilingVectors{
+      {tileShapeM}, {tileShapeN}, {tileShapeM, tileShapeN}};
+
+  for (int i = 1; i <= 3; i++) {
+    auto operand = brgemmOp.getOperand(i).getDefiningOp()->getOperand(0);
+    auto operandType = operand.getType();
+    auto resultType = dyn_cast<MemRefType>(brgemmOp.getOperand(i).getType());
+    auto reassociationIndex =
+        getReassociationIndices(originalShapes[i - 1], tilingVectors[i - 1]);
+
+    SmallVector<int64_t> shape;
+    for (size_t j = 0; j < tilingVectors[i - 1].size(); j++) {
+      shape.append(tilingVectors[i - 1][j].begin(),
+                   tilingVectors[i - 1][j].end());
+    }
+    shape.append(
+        std::next(originalShapes[i - 1].begin(), tilingVectors[i - 1].size()),
+        originalShapes[i - 1].end());
+    insertSubview(shape, operandType, resultType, reassociationIndex, operand,
+                  op, b, brgemmOp, i);
+  }
+
+  return op;
+}
+
+bool getInnermostForLoops(Operation *rootOp,
+                          SmallVectorImpl<scf::ForallOp> &result) {
+  assert(rootOp != nullptr && "Root operation must not be a nullptr.");
+  bool rootEnclosesForAllloops = false;
+  for (Region &region : rootOp->getRegions()) {
+    for (Block &block : region.getBlocks()) {
+      for (Operation &op : block) {
+        bool enclosesPloops = getInnermostForLoops(&op, result);
+        rootEnclosesForAllloops |= enclosesPloops;
+        if (auto ploop = dyn_cast<scf::ForallOp>(op)) {
+          rootEnclosesForAllloops = true;
+
+          // Collect forall loop if it is an innermost one.
+          if (!enclosesPloops)
+            result.push_back(ploop);
+        }
+      }
+    }
+  }
+  return rootEnclosesForAllloops;
+}
+
+struct LoopInsertionPass
+    : public tpp::impl::LoopInsertionPassBase<LoopInsertionPass> {
+
+  LoopInsertionPass(){};
+
+  LoopInsertionPass(ArrayRef<unsigned> tileShapeM,
+                    ArrayRef<unsigned> tileShapeN) {
+    this->tileShapeM = tileShapeM;
+    this->tileShapeN = tileShapeN;
+  };
+
+  LoopInsertionPass(const tpp::LoopInsertionPassOptions &options) {
+    tileShapeM = options.tileShapeM;
+    tileShapeN = options.tileShapeN;
+  };
```

Comment:
These are autogenerated from TD.

Reply:
I have removed the arrayref initializers but the options initializer needs to stay because the compiler complains otherwise.

Reply:
You can still remove these but you need to add `using LoopInsertionPassBase::LoopInsertionPassBase;` (or whatever base class name is for others).